warp/src/js_exports/raygun.rs

```diff
@@ -11,3 +23,746 @@ impl RayGunBox {
+        self.inner.created().into()
+    }
+    pub fn modified(&self) -> js_sys::Date {
+        self.inner.modified().into()
+    }
+    pub fn settings(&self) -> JsValue {
+        serde_wasm_bindgen::to_value(&self.inner.settings()).unwrap()
+    }
+    pub fn recipients(&self) -> Vec<String> {
+        self.inner
+            .recipients()
+            .iter()
+            .map(|did| did.to_string())
+            .collect()
+    }
+}
+
+#[wasm_bindgen]
+pub struct Messages {
+    variant: MessagesEnum,
+    value: JsValue,
+}
+#[wasm_bindgen]
+impl Messages {
+    fn new(messages: raygun::Messages) -> Self {
+        match messages {
+            raygun::Messages::List(list) => Self {
+                variant: MessagesEnum::List,
+                value: serde_wasm_bindgen::to_value(
+                    &list
+                        .iter()
+                        .map(|m| Message::new(m.clone()))
+                        .collect::<Vec<Message>>(),
+                )
+                .unwrap(),
+            },
+            raygun::Messages::Stream(stream) => Self {
+                variant: MessagesEnum::List,
+                value: AsyncIterator::new(Box::pin(stream.map(|s| Message::new(s).into()))).into(),
+            },
+            raygun::Messages::Page { pages, total } => {
+                let pages = pages
+                    .iter()
+                    .map(|p| MessagePage {
+                        id: p.id(),
+                        messages: p
+                            .messages()
+                            .iter()
+                            .map(|m| Message::new(m.clone()))
+                            .collect(),
+                        total: p.total(),
+                    })
+                    .collect();
+
+                Self {
+                    variant: MessagesEnum::Page,
+                    value: serde_wasm_bindgen::to_value(&Page { pages, total }).unwrap(),
+                }
+            }
+        }
+    }
+    pub fn variant(&self) -> MessagesEnum {
+        self.variant
+    }
+    pub fn value(&self) -> JsValue {
+        self.value.clone()
+    }
+}
+#[derive(Serialize, Deserialize)]
+pub struct Page {
+    pub pages: Vec<MessagePage>,
+    pub total: usize,
+}
+#[derive(Serialize, Deserialize)]
+pub struct MessagePage {
+    pub id: usize,
+    pub messages: Vec<Message>,
+    pub total: usize,
+}
+#[wasm_bindgen]
+#[derive(Copy, Clone)]
+pub enum MessagesEnum {
+    List,
+    Stream,
+    Page,
+}
+
+#[wasm_bindgen]
+pub struct MessageOptions {
+    inner: raygun::MessageOptions,
+}
+#[wasm_bindgen]
+impl MessageOptions {
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self {
+            inner: raygun::MessageOptions::default(),
+        }
+    }
+
+    pub fn set_date_range(&mut self, range: JsValue) {
+        self.inner = self
+            .inner
+            .clone()
+            .set_date_range(serde_wasm_bindgen::from_value(range).unwrap());
```

Comment:
wouldnt it have been better to mutate `inner` directly instead of cloning inner and mutating that instance?

Reply:
I would have preferred that, but the way `MessageOptions` is implemented, all of its methods move the instance, which is useful for chaining methods, but doesn't really play nice in this case since the data lives on the wasm side and js only holds a reference to it. 

https://github.com/Satellite-im/Warp/blob/bc7a794302092987816b05e99857a1fefeb6d1dc/warp/src/raygun/mod.rs#L133-L140

I do think it would be nice to refactor this is another PR, but I opted for cloning here at the moment to reduce the amount of code to write/change in this PR compared to other alternatives:

Option 1 - Change `MessageOptions` to act on a reference of self instead, and updating all the places throughout the code base where these methods are called. (I prefer this for consistency with the rest of the code)

Option 2 - Create a new struct similar to MessageOptions with to and from conversion functions, using only js compatible types to be able to have it live in js memory and chain method calls from there.

Reply:
We could make that change in a future PR